### ISSUE TITLE
Make all IO asynchronous all the way to the driver

### ISSUE THREAD

**Issue by DemiMarie** [redox-os/redox#866]
All IO in Redox should be asynchronous all the way to the driver.  This includes operations such as creating and deleting files and directories.  Applications should not need to maintain a thread pool for IO operations.

**Comment by Thetoxicarcade** [comment]
can... we keep both behaviors synthetically? Are there downsides to IO being asynchronous all the way through?

**Comment by RobertWHurst** [comment]
This would be extremely cool 👍 
Would love to see this.

**Comment by ewtoombs** [comment]
I'm not convinced ASIO is universally preferable. In my experience, SIO is often simpler to use and under many circumstances IO is instantaneous and reliable enough that the time spent blocking doesn't really matter. Or I don't know. _Is_ ASIO always better and just as simple to write programs for? I'd hate to see redox get sunk over stances that may prove dogmatic and often wrong. (*cough* HURD * cough*)

**Comment by RobertWHurst** [comment]
@ewtoombs sync IO is easier yes, but it's also limiting when you desire concurrency. It also leads to off loading unnecessary complexity onto programs that need concurrency, such as the need to manage threads for IO.

> under many circumstances IO is instantaneous and reliable enough that the time spent blocking doesn't really matter

I believe an OS should aim to provide the most flexible and capable IO interface. An async interface can be easily made sync. The same cannot be said for the reverse. As for performance, it always matters at an OS level. For your use case it may be only a nano seconds difference, but for something like a video encoder it could end up in hours of thread time wasted.

**Comment by ewtoombs** [comment]
> I believe an OS should aim to provide the most flexible and capable IO interface. An async interface can be easily made sync.

An excellent point. Make the interface to the OS as simple as possible, handle the rest with application-level software and libraries. All right, I'm convinced lol.

**Comment by RobertWHurst** [comment]
@ewtoombs I didn't say simple as possible. I said the most flexible and capable.

**Comment by ewtoombs** [comment]
Well whatever. There's no point int inplementing ASIO and SIO if ASIO can be used to implement SIO libraries at the application level. That's what I mean.

**Comment by RobertWHurst** [comment]
@ewtoombs There is a point actually. Offering async I/O reduces the complexity needed to create a program.

**Comment by ticki** [comment]
SIO is a subset of AIO to be exact. You can spawn a task and then wait for it. Ideally, all this can be done in one syscall through bulk calls. So there doesn't have to be two interfaces in practice. For example, `write(FD, buf, n)` becomes `awrite(FD, buf, n); wait(FD);` or something like that.

I haven't gotten around to work on the new kernel so much yet, but my approach would be something like this:

Every resources can have a number of associated tasks (perhaps reduce this to a single task per resource and then use DUP to create multiple tasks). Things like read and write spawns such task and gives some kind of identifier to the task. These tasks all-together has an API like this:

- `chain(a, b)` - appends `b` to `a` such that the new `a` will complete first when the two tasks complete (in order `a`, `b`). `b` is invalidated.

- `join(a, b)` - appends `b` to `a` such that the new `a` will complete first when the two tasks complete (without any order). `b` is invalidated.

- `wait(a)` - blocks until `a` completes and returns the return value of the completed task.

- `status(a)` - returns the return code of `a`, `-1` if not completed.

- `spawn(a)` - spawns `a` (event loop).

Something like that.

Paging experts in async io: @carllerche @alexcrichton

**Comment by ticki** [comment]
@RobertWHurst

> @ewtoombs There is a point actually. Offering async I/O reduces the complexity needed to create most programs. I'd add that offering both sync and async interfaces is likely the cheapest at the OS level because both interfaces can share the same underlying logic.

No, if full AIO is implemented, there is no need for kernel-level SIO. Redox is a microkernel, and microkernels ought to be minimal. If something can be implemented without OS help with proper performance, there is no need for kernel support.

**Comment by Coder206** [comment]
I noticed that the browser is having trouble closing and I heard it needed and asynchronous IO (see [https://github.com/redox-os/redox/issues/806]. Is this the issue I should follow to keep track of the fix?

**Comment by Thetoxicarcade** [comment]
he may just fix it in the browser and then move that fix down the userspace stack, or just go mad-scientist and make it genesis from modifying the kernel API...

mad scientist!

**Comment by jackpot51** [comment]
There is already I/O multiplexing with fevent, it just needs to be expanded upon
